[CI] LuceneQueryExpressionEvaluatorTests testTermQuery failing

**Build Scans:**
- [elasticsearch-periodic #4884 / encryption-at-rest](https://gradle-enterprise.elastic.co/s/v4rgf4fv6nzvq)
- [elasticsearch-pull-request #41176 / part-3](https://gradle-enterprise.elastic.co/s/p3plnrgcajiem)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:compute:test" --tests "org.elasticsearch.compute.lucene.LuceneQueryExpressionEvaluatorTests.testTermQuery" -Dtests.seed=C6352A8C746D6073 -Dtests.locale=twq -Dtests.timezone=Africa/Lubumbashi -Druntime.java=23
```

**Applicable branches:**
main

**Reproduces locally?:**
**yes**

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.compute.lucene.LuceneQueryExpressionEvaluatorTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!(testTermQuery),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
org.elasticsearch.ElasticsearchTimeoutException: java.util.concurrent.TimeoutException: Timeout waiting for task.
```

**Issue Reasons:**
- [main] 2 failures in test testTermQuery (0.5% fail rate in 435 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

The leaf readers don't seem to be happy about docs enums leaving their current thread. This looks like it's related how the test runs queries rather than how `LuceneQueryExpressionEvaluator` works - also, the latter had no changes since July. Thus, marking this as `low-risk`.

```
WARNING: Uncaught exception in thread: Thread[#3743,elasticsearch[LuceneQueryExpressionEvaluatorTests][esql][T#11],5,TGRP-LuceneQueryExpressionEvaluatorTests]	
java.lang.AssertionError: Docs enums are only supposed to be consumed in the thread in which they have been acquired. But was acquired in Thread[#3735,elasticsearch[LuceneQueryExpressionEvaluatorTests][esql][T#3],5,TGRP-LuceneQueryExpressionEvaluatorTests] and consumed in Thread[#3743,elasticsearch[LuceneQueryExpressionEvaluatorTests][esql][T#11],5,TGRP-LuceneQueryExpressionEvaluatorTests].	
	at __randomizedtesting.SeedInfo.seed([C6352A8C746D6073]:0)	
	at org.apache.lucene.tests.index.AssertingLeafReader.assertThread(AssertingLeafReader.java:70)	
	at org.apache.lucene.tests.index.AssertingLeafReader$AssertingPostingsEnum.docID(AssertingLeafReader.java:525)	
	at org.apache.lucene.search.Weight$DefaultBulkScorer.score(Weight.java:258)	
	at org.elasticsearch.compute.lucene.LuceneQueryExpressionEvaluator$SegmentState.scoreDense(LuceneQueryExpressionEvaluator.java:234)	
	at org.elasticsearch.compute.lucene.LuceneQueryExpressionEvaluator.evalSingleSegmentNonDecreasing(LuceneQueryExpressionEvaluator.java:108)	
	at org.elasticsearch.compute.lucene.LuceneQueryExpressionEvaluator.eval(LuceneQueryExpressionEvaluator.java:62)	
	at org.elasticsearch.compute.operator.EvalOperator.process(EvalOperator.java:45)	
	at org.elasticsearch.compute.operator.AbstractPageMappingOperator.getOutput(AbstractPageMappingOperator.java:76)	
	at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:258)	
	at org.elasticsearch.compute.operator.Driver.run(Driver.java:189)	
	at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:378)	
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)	
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:34)	
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1023)	
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)	
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)	
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)	
	at java.base/java.lang.Thread.run(Thread.java:1575)
```

--- a/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneQueryExpressionEvaluator.java
+++ b/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneQueryExpressionEvaluator.java
@@ -191,6 +191,7 @@ public class LuceneQueryExpressionEvaluator implements EvalOperator.ExpressionEv
     private class SegmentState {
         private final Weight weight;
         private final LeafReaderContext ctx;
+
         /**
          * Lazily initialed {@link Scorer} for this. {@code null} here means uninitialized
          * <strong>or</strong> that {@link #noMatch} is true.
@@ -198,10 +199,20 @@ public class LuceneQueryExpressionEvaluator implements EvalOperator.ExpressionEv
         private Scorer scorer;
 
         /**
+         * Thread that initialized the {@link #scorer}.
+         */
+        private Thread scorerThread;
+
+        /**
          * Lazily initialed {@link BulkScorer} for this. {@code null} here means uninitialized
          * <strong>or</strong> that {@link #noMatch} is true.
          */
         private BulkScorer bulkScorer;
+
+        /**
+         * Thread that initialized the {@link #bulkScorer}.
+         */
+        private Thread bulkScorerThread;
 
         /**
          * Set to {@code true} if, in the process of building a {@link Scorer} or {@link BulkScorer},
@@ -223,7 +234,10 @@ public class LuceneQueryExpressionEvaluator implements EvalOperator.ExpressionEv
             if (noMatch) {
                 return blockFactory.newConstantBooleanVector(false, length);
             }
-            if (bulkScorer == null) {
+            if (bulkScorer == null ||  // The bulkScorer wasn't initialized
+                Thread.currentThread() != bulkScorerThread // The bulkScorer was initialized on a different thread
+            ) {
+                bulkScorerThread = Thread.currentThread();
                 bulkScorer = weight.bulkScorer(ctx);
                 if (bulkScorer == null) {
                     noMatch = true;
@@ -257,8 +271,11 @@ public class LuceneQueryExpressionEvaluator implements EvalOperator.ExpressionEv
             if (noMatch) {
                 return;
             }
-            if (scorer == null || scorer.iterator().docID() > minDocId) {
-                // The previous block might have been beyond this one, reset the scorer and try again.
+            if (scorer == null || // Scorer not initialized
+                scorerThread != Thread.currentThread() || // Scorer initialized on a different thread
+                scorer.iterator().docID() > minDocId // The previous block came "after" this one
+            ) {
+                scorerThread = Thread.currentThread();
                 scorer = weight.scorer(ctx);
                 if (scorer == null) {
                     noMatch = true;

